app/ui/browser-modern/views/app.jsx

```diff
@@ -0,0 +1,131 @@
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+import React, { Component, PropTypes } from 'react';
+import PureRenderMixin from 'react-addons-pure-render-mixin';
+import { connect } from 'react-redux';
+
+import Style from '../../shared/style';
+import BrowserWindow from './browser/window';
+
+import * as SharedPropTypes from '../model/shared-prop-types';
+import * as PagesSelectors from '../selectors/pages';
+import * as PageEffects from '../actions/page-effects';
+import WebViewController from '../../shared/util/webview-controller';
+import Deferred from '../../shared/util/deferred';
+
+const APP_STYLE = Style.registerStyle({
+  width: '100%',
+  height: '100%',
+});
+
+class App extends Component {
+  constructor(props) {
+    super(props);
+    this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(this);
+    this.webViewController = new WebViewController(() => this.props.pages);
+    this.addWebViewControllerListeners();
+  }
+
+  componentWillMount() {
+    // All navigations need to be performed after this component mounts/updates,
+    // so that the `pages` accessor on the `webViewController` is up to date.
+    // When the redux store updates causing react to update the components tree,
+    // a descendant component may be considered updated before this one, thus
+    // any operation on the `webViewController` won't work in the descendant
+    // components, since this component's properties won't be updated yet.
+    this.updated = new Deferred();
+  }
+
+  componentDidMount() {
+    this.updated.resolve();
+  }
+
+  componentWillReceiveProps() {
+    this.updated = new Deferred();
+  }
+
+  componentDidUpdate() {
+    this.updated.resolve();
+  }
+
+  addWebViewControllerListeners() {
+    this.webViewController.on('navigate-to', (pageId, location) => {
+      this.props.dispatch(PageEffects.navigatePageTo(document, pageId, location));
+    });
+
+    this.webViewController.on('navigate-back', (pageId) => {
+      this.props.dispatch(PageEffects.navigatePageBack(document, pageId));
+    });
+
+    this.webViewController.on('navigate-forward', (pageId) => {
+      this.props.dispatch(PageEffects.navigatePageForward(document, pageId));
+    });
+
+    this.webViewController.on('navigate-refresh', (pageId) => {
+      this.props.dispatch(PageEffects.navigatePageRefresh(document, pageId));
+    });
+  }
+
+  handlePageMount = (pageId, location) => {
+    // Handle the initial navigation when a new Page component mounts.
+    this.updated.promise.then(() => this.webViewController.navigateTo(pageId, location));
+  }
+
+  handleNavigateBack = pageId => {
+    this.webViewController.navigateBack(pageId);
```

Comment:
All of that being said though, there's not a good reason to still keep webViewController with the new frontend. It made sense in the old one, but not anymore. So this will likely be removed.